calyx/src/passes/static_par_conv.rs

```diff
@@ -0,0 +1,141 @@
+// attribute of the nth stmt in the Seq. Returns None if no such value is available.
+fn cycles_if_seq(s: &ir::Control, index: usize) -> Option<&u64> {
+    match s {
+        ir::Control::Seq(seq) => seq.stmts.get(index).and_then(|stmt| {
+            stmt.get_attributes().and_then(|atts| atts.get("static"))
+        }),
+        _ => panic!("Not a sequence"),
+    }
+}
+
+impl Visitor for StaticParConv {
+    fn finish_par(
+        &mut self,
+        s: &mut ir::Par,
+        _comp: &mut ir::Component,
+        _sigs: &ir::LibrarySignatures,
+        _comps: &[ir::Component],
+    ) -> VisResult {
+        //if the par block is empty we can stop
+        if (*s).stmts.is_empty() {
+            return Ok(Action::Continue);
+        }
+
+        //make sure only have seq stmts
+        if !(*s)
+            .stmts
+            .iter()
+            .all(|x| matches!(x, ir::Control::Seq(_seq)))
+        {
+            return Ok(Action::Continue);
+        }
+
+        //make sure each seq has the same number of stmts
+        let lens_vec =
+            (*s).stmts.iter().map(len_if_seq).collect::<Vec<usize>>();
+        let min_seq_len = *lens_vec
+            .iter()
+            .min()
+            .unwrap_or_else(|| panic!("empty par block"));
+        let max_seq_len = *lens_vec
+            .iter()
+            .max()
+            .unwrap_or_else(|| panic!("empty par block"));
+        if min_seq_len != max_seq_len {
+            return Ok(Action::Continue);
+        }
+        let seq_len = min_seq_len;
+
+        //make sure nth statement in each seq takes same number of cycle
+        for n in 0..seq_len {
+            let cycles_vec = (*s)
+                .stmts
+                .iter()
+                .map(|stmt| cycles_if_seq(stmt, n))
+                .collect::<Vec<Option<&u64>>>();
+            if cycles_vec.is_empty() {
+                panic!("empty par block");
+            }
+            let fst = cycles_vec[0];
+            if fst == None {
+                return Ok(Action::Continue);
+            }
+            if !cycles_vec.into_iter().all(|xth| xth == fst) {
+                return Ok(Action::Continue);
+            }
+        }
+
+        //It complains when I try to use the vec![] syntax. It says:
+        //the trait `Clone` is not implemented for `ir::control::Control`
+        //note: required by a bound in `from_elem`
+        //So I am doing it this way instead.
+        let mut new_seq_stmts = Vec::new();
+        for _n in 0..seq_len {
+            new_seq_stmts.push(Vec::new());
+        }
+
+        for con in s.stmts.drain(..) {
+            match con {
+                ir::Control::Seq(mut seq) => {
+                    for (counter, stmt) in seq.stmts.drain(..).enumerate() {
+                        new_seq_stmts[counter].push(stmt);
+                    }
+                }
+                _ => panic!("Encountered non sequences"),
```

Comment:
Use `unreachable!` instead of `panic!` to signal that this state shouldn't have been reached.